update Secrets docs w/ new custom rules writing features (#1496)
semgrep/semgrep-docs: docs/semgrep-secrets/rules.md

---
slug: rules
append_help_link: true
title: Write rules
hide_title: true
description: Learn about Semgrep Secrets rules.
tags:
  - Semgrep Secrets
---

# Semgrep Secrets rule structure and sample

This article walks you through writing, publishing, and using Semgrep Secrets rules. It also demonstrates what a sample Semgrep Secrets rule looks like, with subsequent sections describing the key-value pairs in the context of a Semgrep Secrets rule.

## Write a rule

There are two ways to write a rule for Semgrep Secrets:

1. Create a YAML file.
2. Use the Semgrep editor.

### Create a YAML file

If you're familiar with Semgrep's rules syntax, including the [validator syntax](/semgrep-secrets/validators), you can create a YAML file containing your rules. When you're done, publish your rules for use with your organization by running the following in the CLI:

```console
semgrep publish my-rule-file.yaml
```

If you want to keep your rules file local, you must pass in the `--allow-untrusted-validators` flag when calling `semgrep ci` from the CLI.

### Use Semgrep Editor

> The Semgrep Editor, available in Semgrep Cloud Platform, can help you write rules for use with Semgrep Secrets. To pull up a sample rule that you can modify:

Reviewer: ```suggestion
The Semgrep Editor, available in Semgrep Cloud Platform, can help you write custom Semgrep Secrets. To pull up a sample rule that you can modify:
```

See [<i class="fa-regular fa-file-lines"></i> Validators](/semgrep-secrets/validators/) for more information.
:::

## Metavariable binding

> Semgrep secrets uses metavariables. Metavariables allow Semgrep Secrets to reuse matched information from your code in its validators. An example of a metavariable is as follows:

Reviewer: ```suggestion
Semgrep Secrets can use metavariables. Metavariables allow Semgrep Secrets to reuse matched information from your code in its validators. An example of a metavariable is as follows:
```

<iframe title="Message displays metavariable content" src="https://semgrep.dev/embed/editor?snippet=JDzRR" width="100%" height="432px" frameBorder="0"></iframe>
<br />

> When you click **Run**, the content from the metavariable `$HELLO` displays as `This content is now reusable in validators`. If this were a secret, Semgrep Secrets could use this to call the appropriate service to determine if the secret is active.

Reviewer: ```suggestion
When you click **Run**, the content from the metavariable `$HELLO` displays as `This content is now reusable in validators`. If this is a Secret rule, Semgrep Secrets could use this to call the appropriate service to determine if the secret is active.
```
Not sure if I understood the intent of the sentence, just tried to make it clear.
